### PR TITLE
Dynamically generate external extension point assembly

### LINKED ISSUE #1094: Generate external assembly with extension points for integration tests
As noted in https://github.com/FakeItEasy/FakeItEasy/pull/1090#issuecomment-300454071, the current assembly will not be compiled if a user runs the integration tests from the IDE without doing a command-line build, and then the tests would fail.

We could generate the assembly as part of the test fixture setup.

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #1094

### CODE REVIEW

File: tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+        /// <summary>
+        /// Gets the path to a generated assembly that contains extension points, but is not referenced by the test assembly.
+        /// </summary>
+        public string AssemblyPath1 { get; }
+
+        /// <summary>
+        /// Gets the path to a copy of <see cref="AssemblyPath1"/>.
+        ///
+        public string AssemblyPath2 { get; }
```
Comment (thomaslevesque) [inline tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs]: Maybe `AssemblyOriginalPath` and `AssemblyCopyPath` instead? The 1 and 2 suffixes aren't very descriptive

File: tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                "FakeItEasy.ExtensionPoints.External",
+                syntaxTrees: new[] { CSharpSyntaxTree.ParseText(assemblyContent) },
+                references: references,
+                options: new CSharpCompilationOptions(outputKind: OutputKind.DynamicallyLinkedLibrary));
+
+            this.TempAssemblyDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+            Directory.CreateDirectory(this.TempAssemblyDirectory);
+
+            this.AssemblyPath1 = Emit(compilation, this.TempAssemblyDirectory, "1");
+            this.AssemblyPath2 = Emit(compilation, this.TempAssemblyDirectory, "2");
```
Comment (thomaslevesque) [inline tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs]: Why generate the assembly a second time? Couldn't you just copy the first file?

File: tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs (hunk 1 above)
Comment (blairconrad) [inline tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs]: Yeah, I went back and forth on this. Will change as you suggest.

File: tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs (hunk 2 above)
Comment (blairconrad) [inline tests/FakeItEasy.IntegrationTests/ExternalAssemblyGenerator.cs]: Too lazy! I'd have to make the directory and then copy.  I was of two minds about this as well. I'll update as you say.

**Comment by blairconrad** [comment]
Okay, changes pushed (as a fixup). I switched to a consistently-named directory, cleaning it up beforehand, because I found that the assembly deletion was failing after the tests were run.
I'm not sure why I didn't see this the first time around.

**Comment by thomaslevesque** [comment]
LGTM, thanks!

I'll let you squash before I merge. I'll need this for testing #1092.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. Squashed!

**Comment by adamralph** [comment]
Moving the milestone from here to #1094

**Pull request merged**